gadget/mountedfilesystem.go

```diff
@@ -0,0 +1,158 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type MountedFilesystemWriter struct {
+	rootDir string
+	ps      *PositionedStructure
+}
+
+// NewMountedFilesystemWriter returns a writer capable of deploying provided
+// structure, with content of the structure stored in the given root directory.
+func NewMountedFilesystemWriter(rootDir string, ps *PositionedStructure) (*MountedFilesystemWriter, error) {
+	if ps.IsBare() {
+		return nil, fmt.Errorf("structure %v has no filesystem", ps)
+	}
+	fw := &MountedFilesystemWriter{
+		rootDir: rootDir,
+		ps:      ps,
+	}
+	return fw, nil
+}
+
+func remapPreserve(dstDir string, preserve []string) []string {
+	preserveInDst := make([]string, len(preserve))
+	for i, p := range preserve {
+		preserveInDst[i] = filepath.Join(dstDir, p)
+	}
+	sort.Strings(preserveInDst)
+
+	return preserveInDst
+}
+
+// Deploy deploys structure data into provided directory. All existing files are
```

Comment:
why "deploy"?   the struct is called "*writer",  I would use either "write" or "install"

Reply:
Changed deploy -> write in relevant places.